sql: panic: runtime error: index out of range

Assigning for triage because this looks like a stats bug? And sadly I wasn't able to figure out exactly which SQL statement caused this. You may have to look in the logs of the linked issue.

```
panic: runtime error: index out of range

goroutine 1220765 [running]:
github.com/cockroachdb/cockroach/pkg/sql/stats.(*SampleReservoir).SampleRow(0xc4250d1978, 0xc4247d2848, 0x0, 0x0, 0x6526df4030f336f6, 0x2605c60, 0xc42fd39b18)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/stats/row_sampling.go:97 +0x40d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*samplerProcessor).mainLoop(0xc4250d1500, 0x2cbc060, 0xc4227c1d80, 0xc42031a260, 0x7, 0x2cbc060)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sampler.go:202 +0x329
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*samplerProcessor).Run(0xc4250d1500, 0x2cbc060, 0xc4227c1d80, 0xc425fecf88)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sampler.go:153 +0x115
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:569 +0x308
```

https://teamcity.cockroachdb.com/viewLog.html?buildId=1058062&buildTypeId=Cockroach_Nightlies_RandomSyntaxTests

Found with RSG.

I can reproduce as follows:
```
> set sql_safe_updates=false;
> create table t (x int);
> insert into t values (1);
> alter table t drop column x;
> create statistics s from t;
panic: runtime error: index out of range

goroutine 772 [running]:
github.com/cockroachdb/cockroach/pkg/sql/stats.(*SampleReservoir).SampleRow(0xc421c5a478, 0xc421515048, 0x0, 0x0, 0x6dbc20c783f9fb0f, 0x678cdc0, 0xc421c65b18)
	/Users/becca/go/src/github.com/cockroachdb/cockroach/pkg/sql/stats/row_sampling.go:97 +0x40d
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*samplerProcessor).mainLoop(0xc421c5a000, 0x6f27800, 0xc4215b4d40, 0xc42037f5e0, 0x7, 0x6f27800)
	/Users/becca/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sampler.go:202 +0x329
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*samplerProcessor).Run(0xc421c5a000, 0x6f27800, 0xc4215b4d40, 0xc420ec1308)
	/Users/becca/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/sampler.go:153 +0x115
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal
	/Users/becca/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:569 +0x308
```

--- a/pkg/sql/create_stats.go
+++ b/pkg/sql/create_stats.go
@@ -132,6 +132,11 @@ func createStatsDefaultColumns(
 		}
 	}
 
+	// If there are still no columns, return an error.
+	if len(pn.columns) == 0 {
+		return nil, errors.New("CREATE STATISTICS called on a table with no visible columns")
+	}
+
 	return pn, nil
 }
 

--- a/pkg/sql/logictest/testdata/logic_test/distsql_stats
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_stats
@@ -253,3 +253,11 @@ WHERE statistics_name = '__auto__'
 column_names  row_count  distinct_count  null_count
 {a}           10000      10              0
 {b}           10000      10              0
+
+# Regression test for #33195.
+statement ok
+CREATE TABLE t (x int); INSERT INTO t VALUES (1); ALTER TABLE t DROP COLUMN x
+
+# Ensure that creating stats on a table with no columns does not cause a panic.
+statement error pq: CREATE STATISTICS called on a table with no visible columns
+CREATE STATISTICS s FROM t

